classes/Tags/TagRepository.php

```diff
@@ -0,0 +1,516 @@
+use Psr\Log\LoggerInterface;
+use User;
+
+class TagRepository
+{
+	const TABLE_NAME = 'thankyou_tag';
+
+	const TAGGING_TABLE = 'thankyou_tagged';
+
+	private $db;
+
+	protected $logger;
+
+	private $query_factory;
+
+	private $tag_factory;
+
+	public function __construct(DbInterface $db, QueryFactory $query_factory, LoggerInterface $logger, TagFactory $tag_factory)
+	{
+		$this->db            = $db;
+		$this->logger        = $logger;
+		$this->query_factory = $query_factory;
+		$this->tag_factory   = $tag_factory;
+	}
+
+	/**
+	 * Returns an array of Tags indexed by their ID.
+	 *
+	 * @param int[] $ids
+	 * @return Tag[]
+	 */
+	public function GetTags(array $ids): array
+	{
+		if (count($ids) === 0)
+		{
+			return [];
+		}
+
+		foreach ($ids as $id)
+		{
+			if (!is_int($id))
+			{
+				throw new InvalidArgumentException("Failed to Load Tags from Database, Tag IDs must be integers");
+			}
+		}
+
+		$query   = "SELECT * FROM " . self::TABLE_NAME . " WHERE id IN in:int:ids";
+		$results = $this->db->query($query, $ids);
+
+		return $this->GetTagsFromDbQuery($results);
+	}
+
+	/**
+	 * @param int|null    $limit
+	 * @param int|null    $offset
+	 * @param string|null $name
+	 * @param array|null  $orders
+	 * @return Tag[]
+	 */
+	public function GetFilteredTags(?int $limit = null, ?int $offset = null, ?string $name = null, ?array $orders = null): array
+	{
+		$query_string = "SELECT * FROM " . self::TABLE_NAME;
+		if (isset($orders))
+		{
+			$query_string .= $this->QueryBuildOrderString($orders);
+		}
+
+		$query = new QueryBuilder($query_string);
+		if (isset($name))
+		{
+			$query->AddSubstringFilter('name', $name);
+		}
+
+		$query->setLimit($limit, $offset);
+
+		$results = $this->db->query($query->GetQuery());
+
+		return $this->GetTagsFromDbQuery($results);
+	}
+
+	/**
+	 * Returns the total number of Tags in the Database.
+	 *
+	 * @return int
+	 */
+	public function GetTotalTags(): int
+	{
+		return (int) $this->db->query_row("SELECT COUNT(1) FROM " . self::TABLE_NAME)[0];
+	}
+
+	/**
+	 * Filters Tags and returns an array of the number of Taggings they have, indexed by their IDs.
+	 *
+	 * @param int|null   $limit
+	 * @param int|null   $offset
+	 * @param bool|null  $active
+	 * @param array|null $orders
+	 * @return array
+	 */
+	public function GetTagsTaggingTotals(?int $limit = null, ?int $offset = null, ?bool $active = null, ?array $orders = null): array
+	{
+		$query_string = "SELECT COUNT(" . self::TAGGING_TABLE . ".item_id) AS total, " . self::TAGGING_TABLE . ".tag_id FROM " . self::TAGGING_TABLE . " GROUP BY " . self::TAGGING_TABLE . ".tag_id";
+
+		if (isset($orders))
+		{
+			$query_string .= $this->QueryBuildOrderString($orders);
+		}
+
+		$query = new QueryBuilder($query_string);
+		$query->AddJoin(self::TAGGING_TABLE, self::TABLE_NAME, self::TABLE_NAME, self::TABLE_NAME . ".id = " . self::TAGGING_TABLE . ".tag_id");
+
+		if (isset($active))
+		{
+			$query->AddWhereAndClause("active = " . (int) $active);
```

Comment:
Be sure to use query parameters instead:

```php
$query->AddWhereAndClause("active = int:active", (int) $active);
```

Reply:
Fixed in cc02f4f1a006488a1e66d969ad0b61b4267e90ea